pkg/arn/arn.go

```diff
@@ -0,0 +1,72 @@
+package arn
+
+import (
+	"fmt"
+	"strings"
+
+	awsarn "github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/aws/aws-sdk-go/aws/endpoints"
+)
+
+// Canonicalize validates IAM resources are appropriate for the authenticator
+// and converts STS assumed roles into the IAM role resource.
+//
+// Supported IAM resources are:
+//   * AWS account: arn:aws:iam::123456789012:root
+//   * IAM user: arn:aws:iam::123456789012:user/Bob
+//   * IAM role: arn:aws:iam::123456789012:role/S3Access
+//   * IAM Assumed role: arn:aws:sts::123456789012:assumed-role/Accounting-Role/Mary (converted to IAM role)
+//   * Federated user: arn:aws:sts::123456789012:federated-user/Bob
+func Canonicalize(arn string) (string, error) {
+	parsed, err := awsarn.Parse(arn)
+	if err != nil {
+		return "", fmt.Errorf("arn '%s' is invalid: '%v'", arn, err)
+	}
+
+	if err := checkPartition(parsed.Partition); err != nil {
+		return "", fmt.Errorf("arn '%s' does not have a recognized partition", arn)
+	}
+
+	parts := strings.Split(parsed.Resource, "/")
+	resource := parts[0]
+
+	switch parsed.Service {
+	case "sts":
+		switch resource {
+		case "federated-user":
```

Comment:
I think we discussed this somewhat, but I just wanted to make sure that we are ok that federated-user is left as an `sts` identity even though assumed-role is translated into an `iam` identity.  Thoughts?  It seems confusing from a UX standpoint.

Reply:
Federated user only appears as an sts value.  The role is an IAM identity when it is assumed it has the sts value with assumed role.

So these are legal:
* `arn:aws:iam::ACCOUNT_NUMBER:role/RoleName`
* `arn:aws:sts::ACCOUNT_NUMBER:assumed-role/RoleName/Session`
* `arn:aws:sts::ACCOUNT_NUMBER:federated-user/User`

These are not:
* `arn:aws:sts::ACCOUNT_NUMBER:role/RoleName`
* `arn:aws:iam::ACCOUNT_NUMBER:assumed-role/RoleName/Session`
* `arn:aws:iam::ACCOUNT_NUMBER:federated-user/User`

Esentially what we are saying is that with this change you never give a specific assumed-role permissions in the cluster, only users, roles, federated users and root accounts.

Reply:
Ideally, I'd prefer if the mappings contain ARN of IAM identities only and not the other IAM resources i.e. those that are returned by GetCallerIdentity response. Namely
- arn:aws:sts::ACCOUNT_NUMBER:assumed-role/RoleName/Session
- arn:aws:sts::ACCOUNT_NUMBER:assumed-role/RoleName/*
- arn:aws:sts::ACCOUNT_NUMBER:federated-user/User
- arn:aws:iam::123456789012:root
- arn:aws:iam::123456789012:user/Bob

'arn:aws:iam::123456789012:role/RoleName' is not an IAM identity ARN instead it's a resource ARN

Reply:
Since this would change the current config format for users (they specify the role ARN), I'd prefer to change this in a V2.